### PR TITLE
added method 'executeTasks' for execute batch without preimage

### PR DESCRIPTION
**Pull request by andreevym**
Added chaincode method "executeTasks" for execute group of requests like "preimage and batch" but without "preimage".

Use Case:

First user make request (Tx method and args) on backend, and backend cache this request on self side
Second user make request (Tx method and args) on backend cache this request on self side
On event backend get requests from cache and send invoke "executeTasks" with methods and args, each task contains their own "id"
Added mock function for test execute task

### CODE REVIEW

File: core/batcher.go [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+	"strings"
+	"time"
+
+	"github.com/anoideaopen/foundation/core/cachestub"
+	"github.com/anoideaopen/foundation/core/telemetry"
+	"github.com/anoideaopen/foundation/core/types"
+	"github.com/anoideaopen/foundation/hlfcreator"
+	"github.com/anoideaopen/foundation/internal/config"
+	"github.com/anoideaopen/foundation/proto"
+	pb "github.com/golang/protobuf/proto" //nolint:staticcheck
```
Comment (muzykantov) [inline core/batcher.go:18]: I would recommend putting these changes in a separate package. Can we use the latest version of the protobuf library?

File: core/batcher.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+
+const (
+	BatcherBatchExecuteEvent                    = "batcherBatchExecute"
+	TxBatcherRequestType     BatcherRequestType = "tx"
+)
+
+type (
+	BatcherBatchExecuteRequestDTO struct {
+		Requests []BatcherRequest `json:"requests"`
+	}
```
Comment (muzykantov) [inline core/batcher.go:33]: The word "Batcher" is used twice in the name, creating redundancy. Typically, one mention is enough to indicate that it relates to batch operations. The term "DTO" (Data Transfer Object) is also redundant in the Go context, where structs are assumed to be for data transfer.

```
type BatchExecuteRequest struct {
	Requests []BatchRequest `json:"requests"`
}
```

Please follow Go naming conventions for better readability. 

https://google.github.io/styleguide/go/

File: core/batcher.go [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+		BatcherRequestType BatcherRequestType `json:"batcher_request_type"` // BatcherRequestType for choose method to handling request. for example tx, swaps, swaps_keys, multi_swaps, multi_swaps_keys
+	}
+
+	Batcher struct {
+		batchCacheStub *cachestub.BatchCacheStub
+		cc             *ChainCode
+		cfgBytes       []byte
+		ski            string
+		tracingHandler *telemetry.TracingHandler
+	}
```
Comment (muzykantov) [inline core/batcher.go:52]: All fields in the Batcher struct are private, and it’s unclear where the methods are because the code is in a large type block. We need to declare this either along with the methods or make the fields public.

https://google.github.io/styleguide/go/best-practices

File: core/batcher.go [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+	}()
+
+	if len(arguments) != 1 {
+		return nil, fmt.Errorf("expected 1 argument, got %d", len(arguments))
+	}
+
+	var batchDTO BatcherBatchExecuteRequestDTO
+	if err := json.Unmarshal([]byte(arguments[0]), &batchDTO); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal BatcherBatchExecuteRequestDTO: %w", err)
+	}
```
Comment (muzykantov) [inline core/batcher.go:84]: var batchRequest BatchExecuteRequest

https://google.github.io/styleguide/go/decisions#variable-names

File: core/batcher.go [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+		logger.Errorf("Couldn't marshal batch response %s: %s", batchID, err.Error())
+		span.SetStatus(codes.Error, "marshalling batch response failed")
+
+		return nil, fmt.Errorf("failed marshalling batch response %s: %w", batchID, err)
+	}
+
+	return data, nil
+}
+
+const skiPattern = "^[0-9a-f]+$"
```
Comment (muzykantov) [inline core/batcher.go:125]: The `const skiPattern = "^[0-9a-f]+$"` is only used in the `validateBatcherSKI` function. Why is it in the global scope?

File: core/batcher.go [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+	batchCacheStub := cachestub.NewBatchCacheStub(stub)
+
+	batcherSKI := contractCfg.GetBatcherSKI()
+	err = validateBatcherSKI(batcherSKI)
+	if err != nil {
+		return Batcher{}, fmt.Errorf("failed to validate batcher SKI: %w", err)
+	}
+
+	return Batcher{
```
Comment (muzykantov) [inline core/batcher.go:154]: Why is a copy being returned here with return Batcher when all the methods have a pointer receiver?

File: core/batcher.go [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+		}
+		batchResponse.TxResponses = append(batchResponse.TxResponses, txResponse)
+		batchEvent.Events = append(batchEvent.Events, txEvent)
+	}
+
+	if err := b.batchCacheStub.Commit(); err != nil {
+		return nil, nil, fmt.Errorf("failed to commit changes using batchCacheStub: %w", err)
+	}
+
+	return &batchResponse, &batchEvent, nil
```
Comment (muzykantov) [inline core/batcher.go:211]: Why is the code working with a copy and then returning a pointer with return &batchResponse, &batchEvent, nil? Wouldn't it make more sense to allocate the data on the heap right away and use a pointer? What is the rationale?

File: core/batcher.go [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+func (b *Batcher) validatedTxSenderMethodAndArgs(
+	stub *cachestub.BatchCacheStub,
```
Comment (muzykantov) [inline core/batcher.go:215]: The name *cachestub.BatchCacheStub does not conform to Go style due to duplication of the package name in the type name. In Go, it is recommended to avoid redundancy by not repeating the package name in the type name.

Variants:
```
package cachestub

type BatchCache struct {
}
```

or

```
package cachestub

type Batch struct {
}
```

`cachestub.BatchCache` or `cachestub.Batch`

like `bytes.Buffer` etc.

https://google.github.io/styleguide/go/decisions#repetitive-with-package

File: core/batcher.go (hunk 1 above)
Comment (andreevym) [reply core/batcher.go:18]: I can't put batcher to separate pkg because after that I catch circular dependency problems.
And in my task "batcher" not enough time to make refactoring everething needed.

By this reason I use the same code part like already used 69 in foundation library like:

```
pb "github.com/golang/protobuf/proto" //nolint:staticcheck
```

File: core/batcher.go (hunk 2 above)
Comment (andreevym) [reply core/batcher.go:33]: Okay, I absolutely agree about DTO in naming variable, but I want to discuss about the suggested name 'BatchExecute Request' of that struct:

The "Batcher" service is service which will have group of methods.

'Batcher' prefix is the name of service and 'batchExecute' is the name of exists method check batcher will "overloading" (like java style).

Can I call method 'batcherExecute' and struct 'BatcherExecuteRequest'?

Will we see that the method 'batcherExecute' is not the same like 'batchExecute'

Comment (andreevym) [reply core/batcher.go:33]: we already have method name 'batchExecute' for robot and we can't call the current struct 'BatchExecuteRequest'

File: core/batcher.go (hunk 1 above)
Comment (muzykantov) [reply core/batcher.go:18]: The `batcher` clearly belongs in a separate package, following Go's guidelines for modularity and separation of concerns. However, encountering circular dependencies when moving `batcher` suggests a design issue. According to Go best practices, circular dependencies indicate that the modules are too tightly coupled, which complicates the codebase. To adhere to Go's style and best practices, we should refactor to eliminate these dependencies, ensuring a cleaner and more maintainable architecture.

File: core/batcher.go (hunk 2 above)
Comment (muzykantov) [reply core/batcher.go:33]: Go does not support method overloading like Java, so using batcherExecute is not idiomatic. I recommend renaming the method to ExecuteBatch and the struct to ExecuteBatchRequest. If naming conflicts arise, consider moving the logic to a separate package.

https://go.dev/doc/effective_go#names
https://go.dev/wiki/CodeReviewComments#package-comments

Comment (andreevym) [reply core/batcher.go:33]: I have been used word "overloading" to give you information that is the two method with the same business functionality but with different method arguments. I clearly undestood that go and java it's two different lang, but "overloading" usually short word to explain the idea.

That is the reason why i put this word in quotes.

Comment (andreevym) [reply core/batcher.go:33]: I like the names (method ExecuteBatch, struct ExecuteBatchRequest) from your suggestion i will rename the method and structure asap

Comment (andreevym) [reply core/batcher.go:36]: @muzykantov I did it, thank you

File: core/batcher.go (hunk 3 above)
Comment (andreevym) [reply core/batcher.go:53]: @muzykantov I did it, thank you

File: core/batcher.go (hunk 4 above)
Comment (andreevym) [reply core/batcher.go:90]: @muzykantov I did it, thank you

File: core/batcher.go (hunk 5 above)
Comment (andreevym) [reply core/batcher.go]: @muzykantov done, validation SKI removed

File: core/batcher.go (hunk 8 above)
Comment (andreevym) [reply core/batcher.go]: I created the issue for that https://github.com/anoideaopen/foundation/issues/47 because this changes independent with this PR

File: core/batcher.go (hunk 7 above)
Comment (andreevym) [reply core/batcher.go]: @muzykantov I fixed
```
batchResponse := &proto.BatchResponse{}
batchEvent := &proto.BatchEvent{}
```
	
https://github.com/anoideaopen/foundation/pull/45/files#diff-9b00b6c3d2b4b78b314c7cede1b54dc9eceb213a00711ee1229c97f7aed7207dR137

**Comment by andreevym** [comment]
@pfi79 please review this PR

**Comment by pfi79** [comment]
@andreevym please, write an integration test.

File: core/core.go [hunk 9]
```diff
@@ -80,6 +80,7 @@ const (
 	CancelCCTransferFrom = "cancelCCTransferFrom"
 	DeleteCCTransferFrom = "deleteCCTransferFrom"
 	CreateIndex          = "createIndex"
+	ExecuteTasks         = "executeTask"
```
Comment (pfi79) [inline core/core.go:83]: Tasks vs Task

File: core/core.go [hunk 10]
```diff
@@ -591,6 +592,23 @@ func (cc *ChainCode) Invoke(stub shim.ChaincodeStubInterface) (r peer.Response)
 			span.SetStatus(codes.Error, errMsg)
 			return shim.Error(errMsg)
 		}
+
+	case ExecuteTasks:
+		bytes, err := TaskExecutorHandler(
```
Comment (pfi79) [inline core/core.go:597]: Tasks vs Task

File: core/core.go (hunk 9 above)
Comment (andreevym) [reply core/core.go]: thanks, fixed

File: core/external_locks.go [hunk 11]
```diff
@@ -460,7 +460,7 @@ func (bc *BaseContract) verifyLockedArgs(
 		return fmt.Errorf("creating admin address: %w", err)
 	}
 
-	// Request verification
```
Comment (pfi79) [inline core/external_locks.go:463]: Why did you change the comments here

File: core/core.go (hunk 10 above)
Comment (andreevym) [reply core/core.go:597]: I prefer to call the service for execution task in the singular 'TaskExecutor' and NOT in the plural 'TasksExecutor'

like UserService is not UsersService

Do you agree?

File: core/external_locks.go (hunk 11 above)
Comment (andreevym) [reply core/external_locks.go]: I fixed, it's my mistake

File: core/core.go (hunk 10 above)
Comment (pfi79) [reply core/core.go:597]: It seems to me that this service is for handling multiple tasks, not just one.

File: core/task_executor.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const ExecuteTasksEvent = "executeTasks"
+
+var ErrTasksNotFound = errors.New("no tasks found")
+
+// ExecuteTaskRequest represents a request to execute a group of tasks.
+type ExecuteTaskRequest struct {
```
Comment (pfi79) [inline core/task_executor.go:28]: Tasks vs Task

Comment (andreevym) [reply core/task_executor.go]: I fixed, you are right, this is tasks ExecuteTasksRequest like method name

cons ExecuteTasks         = "executeTasks"

File: core/core.go (hunk 10 above)
Comment (andreevym) [reply core/core.go]: fixed change handler method name to TasksExecutorHandler

**Review by pfi79 — APPROVED** [review]
